tests/crc_test.c

```diff
@@ -0,0 +1,154 @@
+#include <aws/cal/hash.h>
+#include <aws/common/byte_buf.h>
+#include <aws/testing/aws_test_harness.h>
+
+static const uint8_t DATA_32_ZEROS[32] = {0};
+static const uint32_t KNOWN_CRC32_32_ZEROES = 0x190A55AD;
+static const uint32_t KNOWN_CRC32C_32_ZEROES = 0x8A9136AA;
+
+static const uint8_t DATA_32_VALUES[32] = {0,  1,  2,  3,  4,  5,  6,  7,  8,  9,  10, 11, 12, 13, 14, 15,
+                                           16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31};
+static const uint32_t KNOWN_CRC32_32_VALUES = 0x91267E8A;
+static const uint32_t KNOWN_CRC32C_32_VALUES = 0x46DD794E;
+
+static const uint8_t TEST_VECTOR[] = {'1', '2', '3', '4', '5', '6', '7', '8', '9'};
+static const uint32_t KNOWN_CRC32_TEST_VECTOR = 0xCBF43926;
+static const uint32_t KNOWN_CRC32C_TEST_VECTOR = 0xE3069283;
+
+typedef struct aws_hash *(crc_fn)(struct aws_allocator *);
+
+/* Makes sure that the specified crc function produces the expected results for known input and output*/
+static int s_test_known_crc(
+    const char *func_name,
+    crc_fn *func,
+    const char *data_name,
+    const struct aws_byte_cursor *input,
+    uint32_t expected,
+    struct aws_allocator *allocator) {
+
+    uint8_t len = (uint8_t)input->len;
+    struct aws_hash *crc = func(allocator);
+    aws_hash_update(crc, input);
+    uintptr_t result = (uintptr_t)crc->impl;
+    ASSERT_HEX_EQUALS(expected, result, "%s(%s)", func_name, data_name);
+
+    /* chain the crc computation so 2 calls each operate on about 1/2 of the buffer*/
+    const struct aws_byte_cursor input_first_half = {
+        .len = (int)(len / 2),
+        .ptr = input->ptr,
+    };
+    const struct aws_byte_cursor input_second_half = {
+        .len = (int)(len - len / 2),
+        .ptr = (input->ptr) + (len / 2),
+    };
+    struct aws_hash *crc1 = func(allocator);
+    aws_hash_update(crc1, &input_first_half);
+    aws_hash_update(crc1, &input_second_half);
+    uintptr_t result1 = (uintptr_t)crc1->impl;
+    ASSERT_HEX_EQUALS(expected, result1, "chaining %s(%s)", func_name, data_name);
+
+    struct aws_hash *crc2 = func(allocator);
+    for (size_t i = 0; i < len; ++i) {
+        const struct aws_byte_cursor input_i = {
+            .len = 1,
+            .ptr = input->ptr + i,
+        };
+        aws_hash_update(crc2, &input_i);
+    }
+    uintptr_t result2 = (uintptr_t)crc2->impl;
+    ASSERT_HEX_EQUALS(expected, result2, "one byte at a time %s(%s)", func_name, data_name);
+    aws_hash_destroy(crc);
```

Comment:
before you destroy it, call update() and finalize() again and to check that they fail like you expect

`ASSERT_ERROR(AWS_ERROR_INVALID_STATE, aws_hash_update(...))`